feat: add basic vim navigation to urlbar

adrianord opened the pull request:
Adds the keybinds ctrl+j and ctrl+k to down and up respectively when the urlbar is active for making selections.

This change adds:
- Preference for activating vim navigation for backwards compatibility
- Patch for activating the preference
- IDs in the locales for the preference
  - Localization is all in English, but I am happy to go back and try to translate a few if that helps, just not my strong suit.

Tested on MacOS and Ubuntu (Parallels)

Behavior we see:
- ctrl+j and ctrl+k take precedence when the urlbar is active
- keybinds bound to ctrl+j and ctrl+j still work while the urlbar is *not* active


https://github.com/user-attachments/assets/612c6116-c184-4a9c-879e-122603ea03c2

adrianord committed:
feat: Add basic vim navigation to urlbar

fix: patch

adrianord committed:
chore: add id to each locale

mr-cheffy commented:
Could there please be some unit tests about this to make sure future firefox versions dont break this functionality?

adrianord committed:
test: add for urlbar vim navigation

adrianord committed:
chore: npm run lint:fix

adrianord commented:
@mr-cheffy I've gone ahead and written a unit test for the vim navigation feature.

It's a little crude, so let me know if you have a better way to force multiple possible selections in the URLBar more reliably. What I've done works, but I'm afraid it might be a little brittle, so I'll leave it up to your judgement.

adrianord commented:
Rebased to fix conflicts, is there anything else that needs to be done in this PR before it can get reviewed?

mr-cheffy commented on locales/cs/browser/browser/preferences/zen-preferences.ftl:
```diff
@@ -140,6 +140,8 @@ zen-urlbar-title = Zen adresní řádek
 zen-urlbar-description = Přizpůsobte si adresní řádek podle svých představ
 zen-urlbar-behavior-label = Chování
+zen-urlbar-vim-navigation =
+  .label = Vim Navigation
```
Could you remove this from all of the languages except `en-US`?

mr-cheffy commented on src/browser/components/preferences/zenLooksAndFeel.inc.xhtml:
```diff
@@ -151,6 +151,10 @@
             </menulist>
         </hbox>
+
+      <checkbox id="zenUrlBarVimNavigation"
```
I woudn't add it here to prevent having too much options, id just leave it as an about:config

adrianord replied:
Sure, I'll remove it here and remove the localization since the localization is only used for the checkbox.

mr-cheffy commented on src/zen/tests/urlbar/browser_vim_navigation.js:
```diff
@@ -0,0 +1,41 @@
+  UrlbarTestUtils: 'resource://testing-common/UrlbarTestUtils.sys.mjs',
+});
+
+add_task(async function test_Vim_Navigation() {
```
Looks great! Maybe a test when the pref is `off` just in case?

adrianord committed:
Revert "chore: add id to each locale"

This reverts commit 382a38c20aafd2963a4834372e52b2e355cacd1d.

adrianord committed:
refactor(zenLooksAndFeel): remove vim navigation option

Will keep this in about:config but removing clutter from the Preferences
page

adrianord committed:
test(browser_vim_navigation): add test for when the option is turned off

adrianord committed:
chore(en-US/zen-preferences): cleanup change

adrianord commented:
Updated the PR to remove the option from Preferences to keep it just in about:config, removed the localization since that was only used for the Preferences checkbox, and added the test for when the about:config option is turned off.

adrianord commented:
Is there anything else that needs to be done to get this feature out?

mr-cheffy committed:
Merge branch 'dev' into urlbar-vim-navigation

mr-cheffy commented on src/browser/components/urlbar/UrlbarController-sys-mjs.patch:
```diff
@@ -2,6 +2,27 @@ diff --git a/browser/components/urlbar/UrlbarController.sys.mjs b/browser/compon
+
++    // Handle vim navigation bindings.
++    if (
++      Services.prefs.getBoolPref("zen.urlbar.vim-navigation.enabled", false) &&
```
Could you please move this pref below the `(event.key == "j" || event.key == "k")` check? Fetching prefs can be slow, we would really only want to do it if we are 100% certain the combination satisfies the condition

adrianord replied:
That's a great call out, made the change.

adrianord committed:
fix(UrlbarController-sys-mjs): move preference check under key event check

adrianord commented:
I see that a recent zen update upgraded to firefox 143.0.4, so I'm building locally to make sure everything in this PR still works.

adrianord commented:
Everything works, ready for re-review

mr-cheffy reviewed (changes requested):
Im seeing tests are failing, did you give these a try?

<img width="985" height="453" alt="image" src="https://github.com/user-attachments/assets/3fd439b3-920e-461a-8748-71894647711d" />

adrianord commented:
I did a full clean of my repo and rebuild then test and I don't see the test failure. What platform are you running on so I can try to reproduce? I ran the tests today on MacOS 26.0.1 and get one failing test that is unrelated to the PR.

<img width="1082" height="179" alt="image" src="https://github.com/user-attachments/assets/d66ce0fb-4b28-4e61-bb5e-569fe28256fb" />

adrianord committed:
chore(*): npm run lint

adrianord commented:
I was able to reproduce the test failure on Ubuntu, will let you know when it is fixed.

KennethHoff commented:
Standard vim key bindings for navigation - assuming `j` and `k` (Without modifiers) is unavailable - is `ctrl-n` and `ctrl-p` (I assume they stand for `next` and `previous` respectively, but not sure about that). `ctrl-j` and `ctrl-k` is unorthodox.

adrianord commented:
@KennethHoff Huh, I hadn't realized this because I'm so used to using `ctrl+j` and `ctrl+k` from my own configuration for so many years. Regardless, I feel like there is a quite a few people who are looking for this either because they use this in their own (neo)vim configuration or because they are coming from Arc where this is standard.

I'll re-think the feature, maybe just give the preference a different name, or dive deeper to make this more configurable.

yatinlala commented:
#10753 has added CTRL-n and CTRL-p by default, though its not customizable. Perhaps this PR should be closed.

adrianord commented:
Will open another PR once I get around to working on making those keys customizable
